splitgraph/ingestion/sqlite/__init__.py

```diff
@@ -0,0 +1,177 @@
+import contextlib
+import itertools
+import os
+import sqlite3
+import tempfile
+from collections.abc import Generator
+from contextlib import contextmanager
+from typing import TYPE_CHECKING, Any, Dict, List, Optional, Tuple, cast
+
+import requests
+from psycopg2.sql import SQL, Identifier
+
+from splitgraph.core.types import (
+    Credentials,
+    IntrospectionResult,
+    Params,
+    TableColumn,
+    TableInfo,
+    TableParams,
+)
+from splitgraph.hooks.data_source.base import LoadableDataSource
+
+if TYPE_CHECKING:
+    from splitgraph.engine.postgres.engine import PostgresEngine
+
+LIST_TABLES_QUERY = """
+SELECT * FROM
+(SELECT
+	tbl_name
+FROM sqlite_master
+WHERE type='table') t
+JOIN pragma_table_info(tbl_name) s
+ORDER BY 1,2;
+"""
+
+# based on https://stackoverflow.com/a/16696317
+def download_file(url: str, local_fh: tempfile._TemporaryFileWrapper) -> int:
+    total_bytes_written = 0
+    with requests.get(url, stream=True, verify=os.environ["SSL_CERT_FILE"]) as r:
+        r.raise_for_status()
+        for chunk in r.iter_content(chunk_size=8192):
+            total_bytes_written += local_fh.write(chunk)
+    return total_bytes_written
+
+
+@contextmanager
+def minio_file(url: str) -> Generator[str, None, None]:
+    with tempfile.NamedTemporaryFile(mode="wb", delete=False) as local_fh:
+        download_file(url, local_fh)
+    yield local_fh.name
+    os.remove(local_fh.name)
+
+
+def query_connection(
+    con: sqlite3.Connection, sql: str, parameters: Optional[Dict[str, str]] = None
+) -> List[Any]:
+    with contextlib.closing(con.cursor()) as cursor:
+        cursor.execute(sql, parameters or {})
+        return cursor.fetchall()
+
+
+@contextmanager
+def db_from_minio(url: str) -> Generator[sqlite3.Connection, None, None]:
+    with minio_file(url) as f:
+        with contextlib.closing(sqlite3.connect(f)) as con:
+            yield con
+
+
+def sql_quote_str(s: str) -> str:
+    return s.replace("'", "''").replace('"', '""')
```

Comment:
nit, minor: this should be two functions since quoting identifiers works differently from quoting values. For example:

- a table name `my 6'5" table` would be quoted as `"my 6'5"" table"`
- a value `my 6'5" table` would be quoted as `'my 6''5" table'`

This function escapes both at the same time. Given that we only use it when loading data from the SQLite table, you could rename it to `quote_ident` and only escape double quotes (also, I'm sure we have `quote_ident` implemented somewhere already).

Reply:
Yep, good point, changed it to use an imported version.